osu.Game.Rulesets.Osu/Objects/Spinner.cs

```diff
@@ -52,13 +62,16 @@ protected override void ApplyDefaultsToSelf(ControlPointInfo controlPointInfo, I
         {
             base.ApplyDefaultsToSelf(controlPointInfo, difficulty);
 
-            const double maximum_rotations_per_second = 477f / 60f;
+            // The average RPS required over the length of the spinner to clear the spinner.
+            double minRps = IBeatmapDifficultyInfo.DifficultyRange(difficulty.OverallDifficulty, clear_rpm_range) / 60;
+
+            // The RPS required over the length of the spinner to receive full score (all normal + bonus ticks).
+            double maxRps = IBeatmapDifficultyInfo.DifficultyRange(difficulty.OverallDifficulty, complete_rpm_range) / 60;
 
             double secondsDuration = Duration / 1000;
-            double minimumRotationsPerSecond = IBeatmapDifficultyInfo.DifficultyRange(difficulty.OverallDifficulty, 1.5, 2.5, 3.75);
 
-            SpinsRequired = (int)(secondsDuration * minimumRotationsPerSecond);
-            MaximumBonusSpins = (int)((maximum_rotations_per_second - minimumRotationsPerSecond) * secondsDuration) - bonus_spins_gap;
+            SpinsRequired = (int)(minRps * secondsDuration);
+            MaximumBonusSpins = Math.Max(0, (int)(maxRps * secondsDuration) - SpinsRequired - bonus_spins_gap);
```

Comment:
About the only thing that I would want to highlight here code-wise is that due to the way this is calculated, i.e. due to `int` truncations being here, the RPM requirements given in the OP are actually *upper* bounds rather than *lower*. As in, it's not "you need to spin at least this fast to complete", it's more of a "*if* you spin this fast, you are *guaranteed* to clear, but you can get away with spinning slower if you get lucky with rounding".

From the test cases added in this PR:

OD 0:

```
> (3000ms / 4) to 1/min

  (3000 * millisecond) / 4 = 80 min^-1

> (3000ms / (4 + 2 + 6)) to 1/min

  (3000 * millisecond) / (4 + 2 + 6) = 240 min^-1
```
OD 5:

```
> (3000ms / 7) to 1/min

  (3000 * millisecond) / 7 = 140 min^-1

> (3000ms / (7 + 2 + 10)) to 1/min

  (3000 * millisecond) / (7 + 2 + 10) = 380 min^-1
```

OD 10: 

```
> (3000ms / 11) to 1/min

  (3000 * millisecond) / 11 = 220 min^-1

> (3000ms / (11 + 2 + 8)) to 1/min

  (3000 * millisecond) / (11 + 2 + 8) = 420 min^-1
```

All values are _ever so slightly_ lower or equal to than the limits given in OP, due to the rounding.

I'm not saying this is actionable, just that I had a different impression of what this was gonna be from the OP - this is still probably okay, though.

Reply:
~~We could make the spins requirement a floating point. That will definitely work well when backporting to stable.~~

Reply:
Jokes aside, I don't think this could easily be resolved, apart from using half spins and getting a bit more precision?